cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -606,6 +614,107 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
      when (isRollForward ev) $
        sendEKGDirectCounter ekgDirect "cardano.node.metrics.served.header.counter.int"
 
+-- | CdfCounter tracks the number of time a value below 'limit' has been seen.
+newtype CdfCounter (limit :: Nat) = CdfCounter Int64
+
+-- | Estimates the CDF for a specific limit 'l' by counting the number of times
+-- a value 'v' is below the limit.
+cdfCounter :: forall a l.
+               ( Num a, Ord a
+               , KnownNat l)
+            => a -> Int -> Int64 -> STM.TVar (CdfCounter l) -> STM Double
+cdfCounter v !size !step tCdf= do
+    when (v < lim) $
+        STM.modifyTVar' tCdf (\(CdfCounter c) -> CdfCounter $ c + step)
+
+    (CdfCounter cdf) <- STM.readTVar tCdf
+    return $! (fromIntegral cdf / fromIntegral size)
+
+  where
+    lim :: a
+    lim = fromInteger $ natVal (Proxy :: Proxy l)
+
+
+-- Add an observation to the CdfCounter.
+incCdfCounter :: Ord a => Num a => KnownNat l => a -> Int -> STM.TVar (CdfCounter l) -> STM Double
+incCdfCounter v size = cdfCounter v size 1
+
+-- Remove an observation from the CdfCounter.
+decCdfCounter :: Ord a => Num a => KnownNat l => a -> Int -> STM.TVar (CdfCounter l) -> STM Double
+decCdfCounter v size = cdfCounter v size (-1)
+
+traceBlockFetchClientMetrics
+  :: forall blk remotePeer.
+     ( )
+  => Maybe EKGDirect
+  -> STM.TVar (SMap.Map SlotNo NominalDiffTime)
+  -> STM.TVar (CdfCounter 1)
+  -> STM.TVar (CdfCounter 2)
+  -> STM.TVar (CdfCounter 5)
+  -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
+  -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
+traceBlockFetchClientMetrics Nothing _ _ _ _ tracer = tracer
+traceBlockFetchClientMetrics (Just ekgDirect) slotMapVar cdf1sVar cdf2sVar cdf5sVar tracer = Tracer bfTracer
+
+  where
+    incCdfs :: NominalDiffTime -> Int -> STM (Double, Double, Double)
+    incCdfs delay size = do
+      cdf1s <- incCdfCounter delay size cdf1sVar
+      cdf2s <- incCdfCounter delay size cdf2sVar
+      cdf5s <- incCdfCounter delay size cdf5sVar
+      return (cdf1s, cdf2s, cdf5s)
+
+    decCdfs :: NominalDiffTime -> Int -> STM ()
+    decCdfs delay size =
+      decCdfCounter delay size cdf1sVar
+       >> decCdfCounter delay size cdf2sVar
+       >> decCdfCounter delay size cdf5sVar
+       >> return ()
+
+    bfTracer :: TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)) -> IO ()
+    bfTracer e@(TraceLabelPeer _ (CompletedBlockFetch p _ _ _ delay)) = do
+      traceWith tracer e
+      case pointSlot p of
+        Origin -> return () -- Nothing to do.
+        At slotNo -> do
```

Comment:
or `forM_ (pointSlot p) $ \slotNo -> ...`

Reply:
Obscure -- but fun. : -)

Perhaps it's `forM_`'s reference to `for` that hurts its perception by implying non-genericity and overly-imperative baggage -- whereas it's, instead more generic than it appears to be.

Reply:
I don't think it's obscure.  This is the main usage of  `traverse_` and `forM_` for `Maybe` like types.  Though I am fine with the explicit pattern match to.